File: gnome-extension/ai-usage@ai-usage-control/lib/claude.js
/* Claude Code provider: reads claude-usage-helper.py.
 *
 * The helper returns the session window plus, when present, the weekly
 * and model-scoped quotas. The bar tracks the 5-hour session — the
 * shortest window Claude reports.
 */
'use strict';

const { GLib } = imports.gi;
const ExtensionUtils = imports.misc.extensionUtils;
const Me = ExtensionUtils.getCurrentExtension();
const Format = Me.imports.lib.format;

const CREDS_PATH = GLib.build_filenamev([GLib.get_home_dir(), '.claude', '.credentials.json']);

function clamp(value) {
    return Math.max(0, Math.min(100, Number(value) || 0));
}

var provider = {
    id: 'claude',
    title: 'Claude',
    helper: 'claude-usage-helper.py',

    /* The helper reads the OAuth token the claude CLI stores here; without
     * that file there is nothing to show, so the indicator stays away. */
    detect() {
        return GLib.file_test(CREDS_PATH, GLib.FileTest.EXISTS);
    },

    parse(data) {
        if (typeof data.percent !== 'number')
            return null;

        const percent = clamp(data.percent);
        const resetsAt = Format.isoToUnix(data.resets_at);
        const rows = ['Session (5 h): %d%% used, %s left (resets at %s)'.format(
            Math.round(percent),
            resetsAt ? Format.formatRemaining(resetsAt) : '?',
            resetsAt ? Format.formatReset(resetsAt) : '?')];

        if (data.seven_day_percent !== undefined) {
            const weekReset = Format.isoToUnix(data.seven_day_resets_at);
            rows.push('Week: %d%% used, resets %s'.format(
                Math.round(clamp(data.seven_day_percent)),
                weekReset ? Format.formatReset(weekReset) : '?'));
        }

        let suffix = null;
        if (data.model_percent !== undefined) {
            const name = data.model_name || 'model';
            const modelReset = Format.isoToUnix(data.model_resets_at);
            rows.push('%s: %d%% used, resets %s'.format(
                name,
                Math.round(clamp(data.model_percent)),
                modelReset ? Format.formatReset(modelReset) : '?'));
            suffix = ' · %s %d%%'.format(name, Math.round(clamp(data.model_percent)));
        }

        return { percent, resetsAt, stale: false, suffix, rows, note: null };
    },
};

File: ai-usage/ai_usage/providers/claude.py
"""Claude Code usage: the helper's reply turned into schema-v1 windows.

The bar tracks the 5-hour session window — the shortest one Claude reports —
and the weekly and model-scoped quotas follow it in order. This ordering used
to live in the extension's aiusagelib/claude.js.
"""

from .. import format, schema
from . import common

ID = "claude"
TITLE = "Claude"
HELPER = "claude-usage-helper.py"

SESSION_MINUTES = 300
WEEK_MINUTES = 10080


def normalise(raw, fetched_at, now=None):
    """One provider entry from one helper reply."""
    now = fetched_at if now is None else now
    if not raw.get("ok"):
        reason = raw.get("error") or "helper"
        return schema.provider_result(
            state=common.failure_state(reason),
            error=reason,
            next_retry_at=common.next_retry_at(reason, raw.get("retry_after"), now),
        )

    windows = [schema.window(
        "session", "session", "Session (5 h)",
        format.clamp_percent(raw.get("percent")),
        format.iso_to_unix(raw.get("resets_at")),
        minutes=SESSION_MINUTES,
    )]

    if raw.get("seven_day_percent") is not None:
        windows.append(schema.window(
            "weekly_all", "weekly", "Week",
            format.clamp_percent(raw.get("seven_day_percent")),
            format.iso_to_unix(raw.get("seven_day_resets_at")),
            minutes=WEEK_MINUTES,
        ))

    if raw.get("model_percent") is not None:
        name = raw.get("model_name") or "model"
        windows.append(schema.window(
            "model:%s" % name.lower().replace(" ", "-"), "model", name,
            format.clamp_percent(raw.get("model_percent")),
            format.iso_to_unix(raw.get("model_resets_at")),
            minutes=WEEK_MINUTES,
        ))

    return schema.provider_result(
        state="ok", windows=windows, source="usage-endpoint", fetched_at=int(fetched_at))
